source/standards/storing-credentials.html.md.erb

```diff
@@ -0,0 +1,48 @@
+---
+title: How to store credentials
+last_reviewed_on: 2021-08-06
+review_in: 6 months
+---
+
+# <%= current_page.data.title %>
+
+Depending on how you [manage your accounts](/standards/accounts-with-third-parties.html), you, your team and the service you run may have credentials or other secrets that you need to store securely.
+
+## Personal credentials
+
+Examples include login details for:
+
+- GitHub
+- AWS
+- GOV.UK Signon
+
+[Use the password manager built into to your browser in the first instance](https://lock.cmpxchg8b.com/passmgrs.html#conclusion).
+
+If you can't do this for some reason, then it's OK to use a third-party password manager. This could be necessary if your browser has an accessibility issue, or if you work with multiple browsers.
+
+Some of the third-party password managers people use at GDS include:
+
+- **LastPass** - Approved by Information Assurance and available to install in the GDS Self Service app. However, you can only use the free version on one type of device.
+- [**1Password**](https://1password.com/)
+- [**BitWarden**](https://bitwarden.com/) - An open source password manager.
+- [**KeePassXC**](https://keepassxc.org/) -  An offline password store, which you may want to backup somewhere.
+
+**Do not use browser extensions that autofill credentials**, as [these are fundamentally vulnerable to attack](https://lock.cmpxchg8b.com/passmgrs.html#interprocess-communication). If you do use an extension, make sure it's configured so that you manually copy / paste credentials.
+
+## Team credentials
+
+Examples include shared login details for:
+
+- Software repositories e.g. npmjs.org, rubygems.org
+- Admin portals e.g. Fastly, DockerHub
+
+There's no established best practice for storing shared credentials at GDS. Some teams have a "credentials" GitHub repository and use [pass](https://www.passwordstore.org/) to store credentials encrypted with GPG. However, [this is fundamentally insecure](https://latacora.micro.blog/2019/07/16/the-pgp-problem.html) and we strongly encourage you to look around for possible alternatives. **Any GitHub repositories that contain secrets should be [internal or ideally private](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-on-github/about-repository-visibility)** to help limit access by former staff.
```

Comment:
Whilst this is all true, it reads as though we shouldn't use GPG for storing credentials, when the conclusion of the Slack thread was pretty much that it's the best/only approach we have right now (in terms of working with SSH keys uploaded to GitHub). Perhaps a rephrase like this would be better:

```suggestion
There's no established best practice for storing shared credentials at GDS. Some teams have a "credentials" GitHub repository and use [pass](https://www.passwordstore.org/) to store credentials encrypted with GPG, based on people's SSH keys uploaded to GitHub. At the time of writing, there are no clear alternatives, but it is worth considering other approaches as [GPG is fundamentally insecure](https://latacora.micro.blog/2019/07/16/the-pgp-problem.html).

**Any GitHub repositories that contain secrets should be [internal or ideally private](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/creating-a-repository-on-github/about-repository-visibility)** to help limit access by former staff.
```

Reply:
We should have something here saying "use individual accounts where possible". Something like:

> Wherever possible, give each person who needs access to these systems their own account. If possible, use single sign on from Google or GitHub, as this avoids the need for a separate process to add and remove accounts for joiners and leavers. Ensure that multi-factor authentication is enabled wherever possible.

> Sometimes shared accounts are necessary (for example for non-human users, like bots and build systems). In these cases try to minimise the privileges given to the shared account, and store the credentials in a secure system which the team can access. Consider putting a process in place to rotate these credentials periodically (e.g. every 6 months) to limit the potential for misuse by former staff.

> There's no established best practice for storing shared credentials at GDS... etc.

I think "fundamentally insecure" is too strong. GPG is bad, sure. But "fundamentally insecure" implies to me that it's broken at such a fundamental level that anyone can pwn me and get my passwords. Which I don't think is the case.

My biggest gripe with GPG is that it's extremely difficult to use. That has a security impact itself, but I'm more bothered by the time we waste working around obscure bugs, and the impact having to use such difficult tools has on our teams. Do frontend developers need access to your team secrets? Great, now they've got to learn how to debug an inevitably broken GPG setup. What about designers? User researchers? Can they even install GPG on their machines? The tooling ends up excluding and othering people, leading to an unnecessary two-tier situation where only developers can access some systems.

Maybe:

> There's no established best practice for storing shared credentials at GDS. Some teams have a "credentials" GitHub repository and use [pass](https://www.passwordstore.org/) to store credentials encrypted with GPG. However, using GPG in this way creates a very high barrier to entry, and can be costly to maintain. Investigate alternatives before adopting GPG based credential stores for new teams.

... maybe with a footnote where we can properly lay into GPG.

Reply:
> Some teams have a "credentials" GitHub repository and use pass to store credentials encrypted with GPG.

Regarding this specifically: is it worth mentioning the issues with this approach, regarding removing access when the historical versions of the password store may remain accessible?

Reply:
Good points everyone! 

> We should have something here saying "use individual accounts where possible"

We do have [this guidance](https://gds-way.cloudapps.digital/standards/accounts-with-third-parties.html) already, which I'm trying not to duplicate but we could make the link to it more obvious.

How about this?

```
[Follow the guidance for managing team credentials in the fist instance](/standards/accounts-with-third-parties.html).

There's no established best practice for storing shared credentials at GDS. Some teams have a "credentials" GitHub 
repository and use [pass](https://www.passwordstore.org/) to store credentials encrypted with GPG. Make sure you
configure the repo as [internal or ideally private](https://docs.github.com/en/github/creating-cloning-and-archiving-
repositories/creating-a-repository-on-github/about-repository-visibility)** to help prevent accidental disclosure.

Using GPG in this way isn't ideal because:

- We can't revoke access to credentials, unless we change them. Anyone who was given access can still decrypt
 credentials using their local copy of the repo and its commit history.

- It creates a high barrier to entry, especially for people less familiar with Unix tooling. We've found GPG tools are
 [generally difficult to use](https://latacora.micro.blog/2019/07/16/the-pgp-problem.html) and keyservers are unreliable.

Investigate alternatives before adopting GPG based credential stores for new teams.
```

Reply:
That sounds good to me 👍🏻

Reply:
Great, I've applied this in https://github.com/alphagov/gds-way/pull/633/commits/32e21f6a7f0b2e54d3fe6c57ef77237b6999e42b.

I'll resolve this now, but anyone else feel free to comment if you think there's more to change here.